menuconfig.py

```diff
@@ -299,6 +299,142 @@
     "brightpurple": curses.COLOR_MAGENTA + 8,
 }
 
+def rgb_to_6cube(rgb):
+    """
+    Convert each component of a RGB color to its approximate index
+    in the xterm 6cube color palette
+    """
+    # The formula for converting a single RGB value to the 6cube index
+    rto6 = lambda x: 0 if x < 48 else max(1, int(round((x - 55) / 40)))
+
+    return [rto6(x) for x in rgb]
+
+def rgb_from_6cube(c6):
+    """
+    Convert a 6cube index triplet to the corresponding RGB value
+    """
+    # The formula from converting the xterm 6cube index to a RGB value
+    rfrom6 = lambda x: 0 if x == 0 else 55 + x*40
+    return [rfrom6(x) for x in c6]
+
+def rgb_to_gridx(rgb):
+    """
+    Convert a RGB color to its closest xterm grayscale color
+    (indexes from 0-23 represents xterm colros 232-255)
+    """
+    gray = int(round((rgb[0] + rgb[1] + rgb[2]) / 3))
```

Comment:
Never mind... think I follow the logic now.

Is this a somewhat correct description?

    Convert an RGB color to an xterm 256-color grayscale color that would
    represent it well if we assume that it is already close to grayscale. This
    "grayscale candidate" can be compared against the best "color candidate" to
    find a good xterm color to represent the color.

"Closest" confused me, because I think you'd need to do something fancier to find the grayscale color that you perceive as similarly bright.

Reply:
Just realized that this also gives the gray with the shortest Euclidean distance. I should do math stuff more often. :)

Is there any point to truncating the value already in the beginning? Maybe it could be something like this:

```python
# Closest gray (in an Euclidean sense, projecting onto the line r = g = b)
gray = (rgb[0] + rgb[1] + rgb[2])/3

# Closests index in the grayscale palette, which starts at RGB 0x080808, with
# stepping 0x0A0A0A
gray_index = int(round((gray - 8)/10))

# Clamp the index to [0, 23] (corresponding to 232-255)
return max(0, min(gray_index, 23))
```

Reply:
There's a fancier formula for finding the intensity, and it's what I originally used in my prototype:
round(rgb[0] * 0.299 + rgb[1] * 0.587 + rgb[2] * 0.114)

We could still use that, it will probably give better accuracy for matching the shades of gray. I did do a visual comparison with tmux, and seemed a bit better at picking proper gray shades.

Reply:
Source: https://stackoverflow.com/questions/596216/formula-to-determine-brightness-of-rgb-color